anndata/utils.py

```diff
@@ -60,44 +60,92 @@ def convert_to_dict_nonetype(obj: None):
 
 
 @singledispatch
-def dim_len(x, dim):
+def dim_len(x, axis):
     """\
-    Return the size of an array in dimension `dim`.
+    Return the size of an array in dimension `axis`.
 
     Returns None if `x` is an awkward array with variable length in the requested dimension.
     """
-    return x.shape[dim]
+    return x.shape[axis]
 
 
 try:
     from .compat import awkward as ak
 
     @dim_len.register(ak.Array)
-    def dim_len_awkward(x, dim):
-        if dim == 0:
-            # dimension 0 is a special case - it is always of `ArrayType` and has a fixed length.
-            try:
-                return x.type.length
-            except AttributeError:
-                raise ValueError("The outermost type must be an `awkward.Array`!")
+    def dim_len_awkward(array, axis):
+        """Get the length of an awkward array in a given dimension
+
+        Returns None if the dimension is of variable length.
+
+        Code adapted from @jpivarski's solution in https://github.com/scikit-hep/awkward/discussions/1654#discussioncomment-3521574
+        """
+        if axis < 0:  # negative axis is another can of worms... maybe later
+            raise NotImplementedError("Does not support negative axis")
+        elif axis == 0:
+            return len(array)
         else:
-            arr_type = x.type
-            for _ in range(dim):
-                # we need to loop through the nested types for the other dimensions, e.g.
-                # ArrayType(RegularType(ListType(NumpyType('int64')), 200), 100)
-                try:
-                    arr_type = arr_type.content
-                except AttributeError:
-                    # RecordType and UnionType have multiple "contents" entries
-                    raise NotImplementedError(
-                        "This check is currently not implemented for RecordType and UnionType arrays. "
+
+            def size_at_depth(layout, depth, lateral_context, **kwargs):
```

Comment:
Is this a closure? Otherwise I think it'd read better defined outside the function.